File: comms/server/api/pin_set.py
################################################################################
# A set of network pin ranges
################################################################################
# Copyright: IQGeo Limited 2010-2023

from copy import copy
from .pin_range import PinRange


class PinSet(object):
    """
    A set of non-overlapping PinRanges
    """

    # ENH: Better as RangeSet?

    def __init__(self, side, low, high):
        """
        Init slots of self
        """

        self.side = side
        self.ranges = [PinRange(side, low, high)]

    def subtract(self, range):
        """
        A copy of self with RANGE excludes

        Returns a PinSet (or None)"""

        sub_ranges = []

        for self_range in self.ranges:
            for sub_range in self_range.subtract(range):
                sub_ranges.append(sub_range)

        sub_set = copy(self)
        sub_set.ranges = sub_ranges

        return sub_set

    def __str__(self):
        """
        Identifying string for progress messages
        """

        return "{}({})".format(self.__class__.__name__, self.spec)

    @property
    def spec(self):
        """
        Full string representation of self
        """

        range_specs = [r.rangeSpec() for r in self.ranges]

        return "{}-{}".format(self.side, ",".join(range_specs))

File: comms/native/services/api/PinSet.js
//###############################################################################
// A set of network pin ranges
//###############################################################################
// Copyright: IQGeo Limited 2010-2023

import { MywClass } from 'myWorld-base';
import PinRange from './PinRange';

/**
 * A set of non-overlapping PinRanges
 */
class PinSet extends MywClass {
    /**
     * Init slots of self
     */
    constructor(side, low, high) {
        super();
        this.side = side;
        this.ranges = [new PinRange(side, low, high)];
    }

    /**
     * A copy of self with RANGE excludes
     *
     * Returns a PinSet (or None)
     */
    subtract(range) {
        const sub_ranges = [];

        for (const self_range of this.ranges) {
            for (const sub_range of self_range.subtract(range)) {
                sub_ranges.push(sub_range);
            }
        }

        const sub_set = new PinSet(this.side, 0, 0);
        sub_set.ranges = sub_ranges;

        return sub_set;
    }

    /**
     * Identifying string for progress messages
     */
    toString() {
        return `PinSet(${this.spec})`;
    }

    //@property
    /**
     * Full string representation of self
     */
    _spec() {
        const range_specs = [];
        for (const r of this.ranges) range_specs.push(r.rangeSpec());

        return `${this.side}-${range_specs.join(',')}`;
    }
}

// -----------------------------------------------------------------------
//                               PROPERTIES
// -----------------------------------------------------------------------

Object.defineProperty(PinSet.prototype, 'spec', {
    /**
     * String representation of self for inclusion in URN
     */
    get() {
        return `${this._spec()}`;
    }
});

export default PinSet;
